### PR TITLE
[core] Fix placement group GPU assignment bug

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

When the new scheduler was implemented, we introduced a bug by assuming that all dynamic resources (including placement group resources) were non-unit resources. This assumption was fine for everything except GPUs. This PR fixes that by allowing dynamic, unit-resources. 

A lot of this came from @clay4444 (especially the test cases!) in #14891 

## Related issue number

#14759 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/raylet/scheduling/cluster_resource_data.cc [hunk 1]
```diff
@@ -94,6 +94,24 @@ TaskRequest ResourceMapToTaskRequest(
   return task_request;
 }
 
+const std::vector<FixedPoint> TaskResourceInstances::Get(
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_resource_data.cc:97]: Return the reference instead?

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc [hunk 2]
```diff
@@ -402,36 +402,41 @@ const NodeResources &ClusterResourceScheduler::GetLocalNodeResources() const {
...
 }
 
-int64_t ClusterResourceScheduler::NumNodes() { return nodes_.size(); }
+int64_t ClusterResourceScheduler::NumNodes() const { return nodes_.size(); }
 
-void ClusterResourceScheduler::AddLocalResource(const std::string &resource_name,
-                                                double resource_total) {
-  string_to_int_map_.Insert(resource_name);
-  int64_t resource_id = string_to_int_map_.Get(resource_name);
+const StringIdMap ClusterResourceScheduler::GetStringIdMap() const {
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:407]: This too. Return the reference instead. 

```suggestion
const StringIdMap &ClusterResourceScheduler::GetStringIdMap() const {
```

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc [hunk 3]
```diff
@@ -402,36 +402,41 @@ const NodeResources &ClusterResourceScheduler::GetLocalNodeResources() const {
...
-    auto &capacity =
-        local_node_it->second.GetMutableLocalView()->custom_resources[resource_id];
-    capacity.available += total;
-    capacity.total += total;
+void ClusterResourceScheduler::AddLocalResourceInstances(
+    const std::string &resource_name, const std::vector<FixedPoint> &instances) {
+  ResourceInstanceCapacities *node_instances;
+  local_resources_.predefined_resources.resize(PredefinedResources_MAX);
+  if (kCPU_ResourceLabel == resource_name) {
+    node_instances = &local_resources_.predefined_resources[CPU];
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:416]: Why do you use `&local_resources_.` here? (Why do you use & in the beginning of this attribute?)

**Review by rkooo567 — APPROVED** [review]

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc (hunk 3 above)
Comment (wuisawesome) [reply src/ray/raylet/scheduling/cluster_resource_scheduler.cc:416]: To get a pointer. In an ideal world, c++ would let us do `ResourceInstanceCapacities &node_instances` then initialize it in the if statements, but this is the next best thing.

**Comment by ericl** [comment]
ASAN test failures (use after free)

**Comment by amogkam** [comment]
@wuisawesome can you add this to the release blocker spreadsheet please (along with ETA and any tests that need to be re-run).

**Comment by wuisawesome** [comment]
Small+large test failure is a CI bug. Looks fine in buildkite. Merging now.

cc @amogkam

**Pull request merged**